### ISSUE TITLE
feature_request(enhancement): submodules support

### ISSUE THREAD

**Issue by Kristinita** [wilvk/githubdl#10]
### 1. Summary

It would be nice, if GitHub Path Downloader will support [**GitHub submodules**](https://blog.github.com/2016-02-01-working-with-submodules/) downloading.

### 2. Argumentation

I use [**Pelican — static site generator**](https://blog.getpelican.com/). [**It haven't normal manager for plugins downloading**](https://github.com/getpelican/pelican-plugins#how-to-use-plugins). At the time, that [**automatically build my site**](https://kristinita.ru/Pages/About-for-developers#site-building):

1. User [**must clone**](https://github.com/getpelican/pelican-plugins#how-to-use-plugins) all big **<https://github.com/getpelican/pelican-plugins>** repository.
1. **Or** I need [**add third-party, not my plugins to my repository**](https://github.com/Kristinita/KristinitaPelican/tree/e0ac486d3add65dca5df303ae9bd29bba523619e/pelican-plugins).

In my opinion, better solution — download specific plugins. But I don't see the easy way to download submodules via GitHub Path Downloader.

### 3. Data

+ `main.py`

```python
"""githubdl folders."""
import githubdl
import yaml

YAMLCONFIG = yaml.load(open('pelicanvariables.yaml', encoding='utf-8'))
[githubdl.dl_dir(
    "https://github.com/getpelican/pelican-plugins",
    plugin,
    "pelican-plugins/" + plugin) for plugin in YAMLCONFIG["variables"]]

```

### 4. Expected behavior

+ [**Online demonstration on Repl.it**](https://repl.it/@Kristinita/KiraGItHubDl-ExpectedDownloading)

**If** `pelicanvariables.yaml`:

```yaml
variables:
- random_article
- slim
```

Output:

```python
Python 3.6.1 (default, Dec 2015, 13:05:11)
[GCC 4.8.2] on linux
2019-01-18 06:15:11,572 - root         - INFO     - Retrieving a list of files for directory: random_article
2019-01-18 06:15:11,573 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /random_article
2019-01-18 06:15:11,573 - root         - INFO     - Requesting file: random_article at url:https://api.github.com/repos/getpelican/pelican-plugins/contents/random_article
2019-01-18 06:15:11,791 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /random_article/Readme.md
2019-01-18 06:15:11,791 - root         - INFO     - Requesting file: random_article/Readme.md at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/random_article/Readme.md
2019-01-18 06:15:12,047 - root         - INFO     - Writing to file: pelican-plugins/random_article/Readme.md
2019-01-18 06:15:12,047 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /random_article/__init__.py
2019-01-18 06:15:12,047 - root         - INFO     - Requesting file: random_article/__init__.py at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/random_article/__init__.py
2019-01-18 06:15:12,257 - root         - INFO     - Writing to file: pelican-plugins/random_article/__init__.py
2019-01-18 06:15:12,261 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /random_article/random_article.py
2019-01-18 06:15:12,263 - root         - INFO     - Requesting file: random_article/random_article.py at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/random_article/random_article.py
2019-01-18 06:15:12,504 - root         - INFO     - Writing to file: pelican-plugins/random_article/random_article.py
2019-01-18 06:15:12,504 - root         - INFO     - Retrieving a list of files for directory: slim
2019-01-18 06:15:12,504 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /slim
2019-01-18 06:15:12,504 - root         - INFO     - Requesting file: slim at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/slim
2019-01-18 06:15:12,707 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /slim/README.md
2019-01-18 06:15:12,707 - root         - INFO     - Requesting file: slim/README.md at url:https://api.github.com/repos/getpelican/pelican-plugins/contents/slim/README.md
2019-01-18 06:15:12,923 - root         - INFO     - Writing to file: pelican-plugins/slim/README.md
2019-01-18 06:15:12,923 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /slim/__init__.py
2019-01-18 06:15:12,923 - root         - INFO     - Requesting file: slim/__init__.py at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/slim/__init__.py
2019-01-18 06:15:13,123 - root         - INFO     - Writing to file: pelican-plugins/slim/__init__.py
2019-01-18 06:15:13,124 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /slim/slim.py
2019-01-18 06:15:13,124 - root         - INFO     - Requesting file: slim/slim.py at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/slim/slim.py
2019-01-18 06:15:13,357 - root         - INFO     - Writing to file: pelican-plugins/slim/slim.py
```

Plugins (folders) `random_article` and `slim` successful download for me to `pelican-plugins` folder.

### 5. Actual behavior

+ [**Online demonstration on Repl.it**](https://repl.it/@Kristinita/KiraGItHubDl-SubmoduleDownloading):

**Else** `pelicanvariables.yaml`:

```yaml
variables:
- replacer
```

```json
Python 3.6.1 (default, Dec 2015, 13:05:11)
[GCC 4.8.2] on linux
2019-01-18 06:12:27,374 - root         - INFO     - Retrieving a list of files for directory: replacer
2019-01-18 06:12:27,375 - root         - INFO     - repo_name: getpelican/pelican-plugins api_path: contents request_string: /replacer
2019-01-18 06:12:27,375 - root         - INFO     - Requesting file: replacer at url: https://api.github.com/repos/getpelican/pelican-plugins/contents/replacer
2019-01-18 06:12:27,568 - root         - CRITICAL - Unable to retrieve list of files from response.
 Exception: 'str' object has no attribute 'get'
 Response: {'name': 'replacer', 'path': 'replacer', 'sha': '7881a1d838b599dd69c8b83a90f00fee96a3ae44', 'size': 0, 'url': 'https://api.github.com/repos/getpelican/pelican-plugins/contents/replacer?ref=master', 'html_url': 'https://github.com/narusemotoki/replacer/tree/7881a1d838b599dd69c8b83a90f00fee96a3ae44', 'git_url': 'https://api.github.com/repos/narusemotoki/replacer/git/trees/7881a1d838b599dd69c8b83a90f00fee96a3ae44', 'download_url': None, 'type': 'submodule', 'submodule_git_url': 'https://github.com/narusemotoki/replacer', '_links': {'self': 'https://api.github.com/repos/getpelican/pelican-plugins/contents/replacer?ref=master', 'git': 'https://api.github.com/repos/narusemotoki/replacer/git/trees/7881a1d838b599dd69c8b83a90f00fee96a3ae44', 'html': 'https://github.com/narusemotoki/replacer/tree/7881a1d838b599dd69c8b83a90f00fee96a3ae44'}}
```

It would be nice, if `'type': 'submodule'`, GitHub Path Downloader download value of `submodule_git_url` key.

Thanks.

**Comment by wilvk** [comment]
Hi @Kristinita , 

Thanks for this feature request. I'll look into it and hopefully implement it soon.

Feel free to raise a PR for it if it's something urgent.

Cheers,
Will

**Comment by wilvk** [comment]
@Kristinita I've added the `-s` option (`submodules=True` for `dl_dir()` ) which reads the relevant `.gitmodules` file and downloads the git submodules as required. This has been released as `0.1.5` via Pip.

Hope this helps.
